### PR TITLE
Feature/authentication : Account creation & True user data display

### PR DESCRIPTION
**Pull request by JanStaszewiczEPFL**
### PR Summary

This pull request enhances the profile and sign-in experience by introducing the following key changes:

**Account Creation on First Sign-In:**

- Added functionality to create an account automatically upon the user’s first sign-in in the `SignIn` screen.

**Display True User Data in Profile Screens:**

- Adjusted both `ProfileView` to display true user data.
- Adjusted both `OtherProfileView` to display true user data.
- Refactored `MainActivity` to adapt to the updated profile screen parameters and ensure proper data flow.
- Updated `AroundYou` screen to pass the user id as a navigation parameters to display true user data in `OtherProfileView`.
- Updated the `Settings` screens to display actual user data on profile cards.

**Tests Implemented**

- Created comprehensive unit tests for ProfileView and OtherProfileView to cover various states, including loading indicators, data display, and user interactions. This required the creation of a `MockProfileRepository` in `MockProfileBackend` to adapt `MockProfileViewModel` to allow the injection of a mock profile and disabling loading during test.
- Adapted Settings screen tests to pass the necessary parameters for displaying true user data on profile cards.
- Modified navigation tests in AroundYou to account for the new profile loading state, ensuring accurate flow and assertions.
- Updated AroundYou screen tests to handle user ID parameters when interacting with profile cards.
- Added tests for account creation during the sign-in process to confirm successful user onboarding.

**Additional Refactoring**

- Made some ProfilesViewModel parameters open to facilitate testing and ensure accessibility for mocks.

**Additional Information**

For some reasons the coverage on new code is counted as 0% on the OtherProfileView.kt file but Jacoco reports good coverage. We hope getting some help from the coachs on this issue that appears for the second time.

Screenshot 2024-11-03 at 19 31 31

### CODE REVIEW

File: app/src/main/java/com/github/se/icebreakrr/ui/sections/shared/DisplayProfile.kt [hunk 1]
```diff
@@ -95,12 +96,19 @@ fun ProfileHeader(
               .aspectRatio(1f) // Keep the aspect ratio 1:1 (height == width) => a square
               .background(Color.LightGray)
               .testTag("profileHeader")) {
+
         // Profile image
-        Image(
-            painter = painterResource(id = R.drawable.turtle),
```
Comment (Kapestian) [inline app/src/main/java/com/github/se/icebreakrr/ui/sections/shared/DisplayProfile.kt:100]: 😢

**Review by Kapestian — CHANGES_REQUESTED** [review]
Your solution seems to work, I left a few (for real it's not much) comments, but there is a big code smell (the hotfix of profile repository) to (un)correct absolutely. Please read them and tell me when you're done. 

Nice work buddy!

**Review by SamSteu — CHANGES_REQUESTED** [review]
Good code in general

Except for the bad coverage but this must be an issue with SonarCloud from your side. Since you have a really good coverage when running Jacoco on local I guess it is fine.

File: app/src/main/java/com/github/se/icebreakrr/ui/sections/Settings.kt [hunk 2]
```diff
@@ -57,8 +70,14 @@ fun SettingsScreen(navigationActions: NavigationActions) {
             Modifier.fillMaxSize().padding(innerPadding).padding(16.dp).verticalScroll(scrollState),
         verticalArrangement = Arrangement.Top,
         horizontalAlignment = Alignment.Start) {
-          ProfileCard(profile = Profile.getMockedProfiles()[0], isSettings = true) {
-            navigationActions.navigateTo(Screen.PROFILE)
+          if (profile != null) {
```
[the remark this answers is not in the document]
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/sections/Settings.kt:73]: There is no loading system implemented inside the cards.

File: app/src/main/java/com/github/se/icebreakrr/ui/sections/AroundYou.kt [hunk 3]
```diff
@@ -81,7 +81,11 @@ fun AroundYouScreen(
                       items(filteredProfiles.value.size) { index ->
                         ProfileCard(
                             profile = filteredProfiles.value[index],
-                            onclick = { navigationActions.navigateTo(Screen.OTHER_PROFILE_VIEW) })
+                            onclick = {
+                              navigationActions.navigateTo(
+                                  Screen.OTHER_PROFILE_VIEW +
+                                      "?userId=${filteredProfiles.value[index].uid}")
```
[the remark this answers is not in the document]
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/sections/AroundYou.kt:87]: It is working like web urls. You can pass navigation params to a screen with the same syntax as in web, you start with "?" to announce your variables and then write `varName=value` to set the value. During navigation the url would be something like `SCREEN_NAME?userId=1234`. You can then retrieve this value in the corresponding screen using the `NavBackStack` class with `val profileId = navBackStackEntry?.arguments?.getString("userId")` (see top of `OtherProfileView.kt`).

File: app/src/main/java/com/github/se/icebreakrr/ui/sections/shared/DisplayProfile.kt (hunk 1 above)
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/sections/shared/DisplayProfile.kt:100]: It's a sad day for IceBreakrr :(

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/SignIn.kt [hunk 4]
```diff
@@ -85,11 +91,46 @@ fun SignInScreen(navigationActions: NavigationActions) {
...
+                  val newProfile =
+                      Profile(
+                          uid = firebaseUser.uid,
+                          name = firebaseUser.displayName ?: "Unknown",
+                          gender = Gender.OTHER,
+                          birthDate = Timestamp.now(),
+                          catchPhrase = "",
+                          description = "")
+
+                  Log.i("SignIn", "Creating new user: ${newProfile.name}")
```
[the remark this answers is not in the document]
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/authentication/SignIn.kt:124]: You are right!

File: app/src/main/java/com/github/se/icebreakrr/ui/authentication/SignIn.kt [hunk 5]
```diff
@@ -85,11 +91,46 @@ fun SignInScreen(navigationActions: NavigationActions) {
...
+                  profilesViewModel.addNewProfile(newProfile)
+                } else { // user already exists
+                  Log.i("SignIn", "Profile exists: ${profile.name}")
```
[the remark this answers is not in the document]
Comment (JanStaszewiczEPFL) [reply app/src/main/java/com/github/se/icebreakrr/ui/authentication/SignIn.kt:127]: You detect the small issues like an eagle! That's some great advice thank's!

**Review by Kapestian — APPROVED** [review]
2nd review: LGTM

**Comment by JanStaszewiczEPFL** [comment]
image
Please review me :(

**Pull request merged**